pilot/pkg/model/cluster_local_test.go

```diff
@@ -142,6 +142,73 @@ func TestIsClusterLocal(t *testing.T) {
 			host:     "s.ns3.svc.cluster.local",
 			expected: false,
 		},
+		{
+			name: "global",
```

Comment:
Can you add a test case with:

*.cluster.local -> set to clusterLocal = true
* -> set to clusterLocal=false

and make sure it works as expected ( echo, echo.testns, echo.testns.svc and echo.testns.svc.cluster.local - return clusterLocal, everything else - return global ).

I think this is the 'right' way to set this up in a safe mode for migrations - so clusterset or explictly defined hostnames in ServiceEntry are treated as global.

Reply:
I think this would require a somewhat more complex change. The logic behind it just now is suffix-based (which is unchanged before this PR), which basically finds the most exact match.

https://github.com/istio/istio/blob/80be4208c56e3dbf87538f4e11a668128e3c6e6c/pilot/pkg/model/config.go#L302-L313

So in the test case suggested above, `echo`, `echo.testns` and`echo.testns.svc` will return **global**, `echo.testns.svc.cluster.local` will return **local**, and everything else (non `*.cluster.local`) will return **global**.

This is fine for standard K8s service discovery sharing as it uses the full service name when checking. Not sure about service entries etc., but the change does now at least allow specific exceptions for most use cases, but agree there is probably scope for improvement.

Reply:
It's fine to remove echo, echo.testns and echo.testns.svc if it's hard to support - but the rest is still useful to test, since it's the safest option ( to have all .cluster.local names local)

I am not entirely sure how this gets called from the main generation - isn't it after normalization ( adding the suffix ) ? I understand this method can't handle .svc and the short names, and doesn't have to if it is called with normalized FQDNs - but I was expecting that end to end it will work as expected, otherwise it wouldn't make sense ( since a very large set of users use the short names ).

Reply:
Agreed, have added a couple of new tests, one for testing a `svc.cluster.local` endpoint while `*` is not local, and `*.cluster.local` is local, and another to test a non `cluster.local` endpoint with the same conditions.

So far (as I'm looking to migrate to multi-cluster) I've only really encountered sharing of `svc.cluster.local` Kubernetes services between the clusters, so not enough knowledge on my side yet regarding sharing of Service Entries etc.!